### PR TITLE
Use goto_programt::make_X factories in goto_convert

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-programs/goto_convert.cpp [hunk 1]
```diff
@@ -531,9 +531,7 @@ void goto_convertt::convert(
   // make sure dest is never empty
   if(dest.instructions.empty())
   {
-    dest.add_instruction(SKIP);
-    dest.instructions.back().code.make_nil();
-    dest.instructions.back().source_location=code.source_location();
+    dest.add(goto_programt::make_skip(code.source_location()));
```
Comment (tautschnig) [inline src/goto-programs/goto_convert.cpp:534]: Looking at the use of these, couldn't we change the approach a bit and make `goto_programt::add` be a templated function taking a variadic number of arguments, with specialisations for the various instruction types?
````c++
template <goto_program_instruction_typet T, typename ... Params>
add(Params...);
template <> add<SKIP, source_locationt> (source_locationt loc) { ... };
````
My rationale is that we might eventually forbid to construct an `instructiont` other than via `.add`, and it also saves us typing `goto_programt::`. The above would then be `dest.add<SKIP>(code.source_location());`

**Review by smowton — APPROVED** [review]

File: src/goto-programs/goto_convert.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-programs/goto_convert.cpp:534]: Yes, I've been tempted by that idea. Will explore.

File: src/goto-programs/goto_program.h [hunk 2]
```diff
@@ -814,6 +814,13 @@ class goto_programt
     return instructiont(code_returnt(), irep_idt(), l, RETURN, nil_exprt(), {});
   }
 
+  static instructiont make_return(
+    code_returnt c,
+    const source_locationt &l = source_locationt::nil())
+  {
+    return instructiont(std::move(c), irep_idt(), l, RETURN, nil_exprt(), {});
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h:821]: We need to make progress on #3126, which removes all these `irep_idt()` in the `make_*` functions.

File: src/goto-programs/goto_convert.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-programs/goto_convert.cpp:534]: It should be ok to merge all the `make_X` work as that does the hard work of collecting individual assignments into a single function call. Possibly moving to `add<T>(...)` is then just text transformation.

**Review by tautschnig — APPROVED** [review]

File: src/goto-programs/goto_program.h (hunk 2 above)
Comment (kroening) [reply src/goto-programs/goto_program.h:821]: Indeed!

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 20a908d41907bf6abc819acb18bef2822c921f79).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99273103

**Pull request merged**